src/main/java/org/apache/mesos/state/StateStoreCache.java

```diff
@@ -0,0 +1,301 @@
+
+    protected Optional<FrameworkID> frameworkId;
+
+    protected Map<String, TaskID> nameToId = new HashMap<>();
+    protected Map<TaskID, TaskInfo> idToTask = new HashMap<>();
+    protected Map<TaskID, TaskStatus> idToStatus = new HashMap<>();
+
+    protected Map<String, byte[]> properties = new HashMap<>();
+
+    /**
+     * Returns a cache instance. To ensure consistency, only one singleton cache instance may exist
+     * in the process at a time. This function may be called multiple times, but only if the same
+     * {@link StateStore} instance is provided each time.
+     */
+    public static StateStore getInstance(StateStore store) {
+        RWLOCK.lock();
+        try {
+            if (instance == null) {
+                instance = new StateStoreCache(store);
+            } else if (instance.store != store) {
+                // Disallow subsequent calls to getInstance() with different instances of StateStore.
+                throw new IllegalStateException(String.format(
+                        "StateStoreCache may only be used against a single instance of StateStore. " +
+                        "got[%s] expected[%s]", store, instance.store));
+            }
+            return instance;
+        } finally {
+            RWLOCK.unlock();
+        }
+    }
+
+    @VisibleForTesting
+    StateStoreCache(StateStore store) throws StateStoreException {
+        this.store = store;
+
+        // Use bulk fetches to initialize cache with underlying storage state:
+        frameworkId = store.fetchFrameworkId();
+        for (TaskInfo task : store.fetchTasks()) {
+            nameToId.put(task.getName(), task.getTaskId());
+            idToTask.put(task.getTaskId(), task);
+        }
+        for (TaskStatus status : store.fetchStatuses()) {
+            idToStatus.put(status.getTaskId(), status);
+        }
+        for (String key : store.fetchPropertyKeys()) {
+            properties.put(key, store.fetchProperty(key));
+        }
+    }
+
+    @Override
+    public void storeFrameworkId(FrameworkID fwkId) throws StateStoreException {
+        RWLOCK.lock();
+        try {
+            store.storeFrameworkId(fwkId);
+            frameworkId = Optional.of(fwkId);
+        } finally {
+            RWLOCK.unlock();
+        }
+    }
+
+    @Override
+    public void clearFrameworkId() throws StateStoreException {
+        RWLOCK.lock();
+        try {
+            store.clearFrameworkId();
+            frameworkId = Optional.empty();
+        } finally {
+            RWLOCK.unlock();
+        }
+    }
+
+    @Override
+    public Optional<FrameworkID> fetchFrameworkId() throws StateStoreException {
+        RLOCK.lock();
+        try {
+            return frameworkId;
+        } finally {
+            RLOCK.unlock();
+        }
+    }
+
+    @Override
+    public void storeTasks(Collection<TaskInfo> tasks) throws StateStoreException {
+        RWLOCK.lock();
+        try {
+            store.storeTasks(tasks);
```

Comment:
Is it possible that storeTasks method persists tasks collection fully or partially and throws an exception ? If yes, then that particular case will make the cache inconsistent.